vignettes/task_plans.Rmd

```diff
@@ -0,0 +1,255 @@
+```{r task_steps}
+# step1 creates recipes like this:
+# WI_prep:
+#   command: prep(I('WI'))
+step1 <- create_task_step(step_name = 'prep')
+
+# step2 creates recipes like this:
+# WI_plot.png:
+#   command: visualize(target_name, 'WI_prep', I('Madison'))
+step2 <- create_task_step(
+  step_name = 'plot',
+  target_name = function(task_name, ...) {
+    sprintf('%s_plot.png', task_name)
+  },
+  command = function(target_name, task_name, ...) {
+    capital <- task_config[task_config$id == task_name, 'capital']
+    sprintf("visualize(target_name, %s_prep, I('%s'))", task_name, capital)
+  })
+
+# step3 creates recipes like this:
+# WI.pdf:
+#   command: report(target_name)
+#   depends:
+#    - WI_plot.png
+step3 <- create_task_step(
+  step_name = 'report',
+  target_name = function(task_name, ...) {
+    sprintf('%s.pdf', task_name)
+  },
+  depends = function(task_name, ...) {
+    sprintf('%s_plot.png', task_name)
+  },
+  command = function(task_name, step_name, ...) {
+    sprintf("report(target_name)", task_name)
+  })
+```
+
+You can now create a task_plan that wraps the tasks and steps together.
+```{r plan}
+task_plan_1 <- create_task_plan(
+  task_config$id,
+  list(step1, step2, step3),
+  final_steps='report',
+  ind_dir='states/log')
+```
+
+You'll need the `task_plan` in that form for two things: to create a remake .yml makefile and to run `loop_tasks`. First, the makefile:
+
+```{r makefile, results='markup'}
+create_task_makefile(
+  task_plan=task_plan_1,
+  makefile='task_plan_1.yml',
+  include=c(),
+  packages=c('ggplot2'),
+  sources=c('script.R'),
+  file_extensions=c('ind'),
+  ind_complete=TRUE)
+```
+
+And here are the contents of task_plan_1.yml:
+
+```{r read_makefile, results='markup', comment='', echo=FALSE}
+cat(readr::read_lines('task_plan_1.yml'), sep='\n')
+```
+
+Note that at the end of each state's task list is a final task, e.g., `states/log/WI.ind`, that we didn't directly specify in the task_plan. These recipes for the `complete` step were generated in `create_task_plan` because we allowed the default argument `add_complete=TRUE`.
+
+Next, use the `task_plan` and the `task_makefile` in `loop_tasks`:
+
+```{r delete1, echo=FALSE}
+unlink('.remake')
+unlink('build')
+```
+
+```{r loop1a}
+loop_tasks(
+  task_plan=task_plan_1,
+  task_makefile='task_plan_1.yml')
+```
+
+## Forcing rebuilds
+
+### Forcing rebuild for one task-step
+
+If we try to `scmake` any of the targets in the task makefile, `remake` correctly recognizes that they're up to date already, and the target is not rebuilt. Instead, `AZ_plot.png` is already `OK`:
+```{r try_rebuild_one}
+scmake('AZ_plot.png', 'task_plan_1.yml')
```

Comment:
scmake does two things differently: it enables a shared remote cache, and it has `remake_file` as its second argument so you don't have to type `remake_file` all the time.

I thiiiink that you could run `remake::make` rather than `scipiper::scmake` if you had absolutely no shared cache in your project, but I wouldn't bet the farm on it until someone tested that theory, and I'm not sure why you'd worry too much about it.

I think this info belongs in a vignette other than task_plans.Rmd. I'll make an issue for creating such a vignette.

Reply:
https://github.com/USGS-R/scipiper/issues/45